foundation-models/openai/src/main/java/com/sap/ai/sdk/foundationmodels/openai/OpenAiTool.java

```diff
@@ -0,0 +1,224 @@
+  @Nonnull
+  public static <InputT> Builder1<InputT> forFunction(@Nonnull final Function<InputT, ?> function) {
+    return inputClass ->
+        name -> {
+          final Function<String, Object> exec =
+              s -> function.apply(deserializeArgument(inputClass, s));
+          final var schema = GENERATOR.generateSchema(inputClass);
+          return new OpenAiTool(name, exec, schema, null, null);
+        };
+  }
+
+  /**
+   * Creates a new OpenAiTool instance with the specified function and input class.
+   *
+   * @param <InputT> the type of the input class.
+   */
+  public interface Builder1<InputT> {
+    /**
+     * Sets the name of the function.
+     *
+     * @param inputClass the class of the input object.
+     * @return a new OpenAiTool instance with the specified function and input class.
+     */
+    @Nonnull
+    Builder2 withArgument(@Nonnull final Class<InputT> inputClass);
+  }
+
+  /** Creates a new OpenAiTool instance with the specified name. */
+  public interface Builder2 {
+    /**
+     * Sets the name of the function.
+     *
+     * @param name the name of the function
+     * @return a new OpenAiTool instance with the specified name
+     */
+    @Nonnull
+    OpenAiTool withName(@Nonnull final String name);
+  }
+
+  @Nullable
+  static <T> T deserializeArgument(@Nonnull final Class<T> cl, @Nonnull final String s) {
+    try {
+      return JACKSON.readValue(s, cl);
+    } catch (JsonProcessingException e) {
+      throw new IllegalArgumentException("Failed to parse JSON string to class " + cl, e);
+    }
+  }
+
+  ChatCompletionTool createChatCompletionTool() {
+    final var schemaMap =
+        OpenAiUtils.getOpenAiObjectMapper()
+            .convertValue(getSchema(), new TypeReference<Map<String, Object>>() {});
+
+    return new ChatCompletionTool()
+        .type(FUNCTION)
+        .function(
+            new FunctionObject()
+                .name(getName())
+                .description(getDescription())
+                .parameters(schemaMap)
+                .strict(getStrict()));
+  }
+
+  private static SchemaGenerator createSchemaGenerator() {
+    final var module =
+        new JacksonModule(
+            JacksonOption.RESPECT_JSONPROPERTY_REQUIRED, JacksonOption.RESPECT_JSONPROPERTY_ORDER);
+    return new SchemaGenerator(
+        new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2020_12, OptionPreset.PLAIN_JSON)
+            .without(Option.SCHEMA_VERSION_INDICATOR)
+            .with(module)
+            .build());
+  }
+
+  /**
+   * Executes the given tool calls with the provided tools and returns the results as a list of
+   * {@link OpenAiToolMessage} containing execution results encoded as JSON string.
+   *
+   * @param tools the list of tools to execute
+   * @param msg the assistant message containing a list of tool calls with arguments
+   * @return The list of tool messages with the results.
+   */
+  @Beta
+  @Nonnull
+  public static List<OpenAiToolMessage> execute(
```

Comment:
```suggestion
  static List<OpenAiToolMessage> execute(
```
Why public?

Reply:
> Why public?

I thought of testing purposes or in case of some weird customization for users.